### PR TITLE
Only vote on menu items created by SonataAdminBundle

### LINKED ISSUE #4254: ChildrenVoter affects non Sonata Menus
#### Sonata packages

```bash
$ composer show sonata-project/*

sonata-project/admin-bundle              3.10.3 The missing Symfony Admin G...
sonata-project/block-bundle              3.2.0  Symfony SonataBlockBundle
sonata-project/cache                     1.0.7  Cache library
sonata-project/core-bundle               3.1.2  Symfony SonataCoreBundle
sonata-project/datagrid-bundle           2.2    Symfony SonataDatagridBundle
sonata-project/doctrine-extensions       1.0.2  Doctrine2 behavioral extens...
sonata-project/doctrine-orm-admin-bundle 3.1.2  Symfony Sonata / Integrate ...
sonata-project/easy-extends-bundle       2.1.10 Symfony SonataEasyExtendsBu...
sonata-project/exporter                  1.7.0  Lightweight Exporter library
sonata-project/google-authenticator      1.0.2  Library to integrate Google...
sonata-project/user-bundle               3.2.0  Symfony SonataUserBundle

```

#### Symfony packages

```bash
$ composer show symfony/*
symfony/assetic-bundle     v2.8.1  Integrates Assetic into Symfony2
symfony/monolog-bundle     2.12.0  Symfony MonologBundle
symfony/phpunit-bridge     v3.2.1  Symfony PHPUnit Bridge
symfony/polyfill-apcu      v1.3.0  Symfony polyfill backporting apcu_* func...
symfony/polyfill-intl-icu  v1.3.0  Symfony polyfill for intl's ICU-related ...
symfony/polyfill-mbstring  v1.3.0  Symfony polyfill for the Mbstring extension
symfony/polyfill-php54     v1.3.0  Symfony polyfill backporting some PHP 5....
symfony/polyfill-php55     v1.3.0  Symfony polyfill backporting some PHP 5....
symfony/polyfill-php56     v1.3.0  Symfony polyfill backporting some PHP 5....
symfony/polyfill-php70     v1.3.0  Symfony polyfill backporting some PHP 7....
symfony/polyfill-util      v1.3.0  Symfony utilities for portability of PHP...
symfony/security-acl       v3.0.0  Symfony Security Component - ACL (Access...
symfony/swiftmailer-bundle v2.4.2  Symfony SwiftmailerBundle
symfony/symfony            v2.8.15 The Symfony PHP framework

```

#### Other

```bash
$ composer show knplabs/*
knplabs/knp-menu        2.2.0 An object oriented menu library
knplabs/knp-menu-bundle 2.1.3 This bundle provides an integration of the Kn...

```

## Subject

The `ChildrenVoter` included as part of `SonataAdminBundle` is affecting other `KnpMenus` in different parts of applications.

Building a breadcrumb as per the `KnpMenuBundle` documentation relies on  which returns `true` for the root node based on the `ChildrenVoter` result.

As per the response from https://github.com/KnpLabs/KnpMenuBundle/issues/331 there doesnt seem to be a way to remove a `Voter` from a specifc `Menu` so is it possible  to restrict the `ChildrenVoter` to only `SonataAdmin` Menus?

More info http://stackoverflow.com/questions/41304270/can-you-make-knpmenu-voters-private

## Steps to reproduce

Call `isCurrent` on a `KnpMenu` root node where the current page is a child node of the root. 

## Expected results
Any parent returns true for `isCurrent` if the `ChildrenVoter` is loaded.

This is correct behavior for `SonataAdmin` menu links but shouldnt be applied to non `SonataAdmin` menus.

## Actual results
Only the current item returns true for non `SonataAdmin` menus.

### PR DESCRIPTION
**Pull request by wouterj**
Closes #4254 

## Changelog

```markdown
### Changed
- Changed `ActiveVoter` and `ChildrenVoter` to only work with menu items having the `SonataAdminBundle` extra set.
```

## Subject

This avoids incorrectly matching items outside the admin dashboard. See #4254 for a more detailed description of what happends when this PR is not applied.

### CODE REVIEW

**Comment by greg0ire** [comment]
> See #4254 for a more detailed description of what happends when this PR is not applied.

TL;DR : the voters put their big fat nose in other people's business.

File: Tests/Command/ExplainAdminCommandTest.php [hunk 1]
```diff
@@ -251,14 +251,16 @@ public function testExecute()
         $commandTester = new CommandTester($command);
         $commandTester->execute(array('command' => $command->getName(), 'admin' => 'acme.admin.foo'));
 
+        file_put_contents(getcwd().'/error.txt', $commandTester->getDisplay(true));
```
Comment (greg0ire) [inline Tests/Command/ExplainAdminCommandTest.php]: What is this file? Will it always be generated? Have you considered using `tempnam` for that?

File: Menu/Matcher/Voter/ActiveVoter.php [hunk 2]
```diff
@@ -26,6 +26,10 @@ class ActiveVoter implements VoterInterface
      */
     public function matchItem(ItemInterface $item)
     {
+        if (!$item->getExtra('SonataAdminBundle', false)) {
+            return;
+        }
```
Comment (greg0ire) [inline Menu/Matcher/Voter/ActiveVoter.php]: Is this extra already there?

**Comment by wouterj** [comment]
Whoa, sorry. Submitted a PR with lots of mistakes. Should be correct now.

**Comment by greg0ire** [comment]
> Whoa, sorry. Submitted a PR with lots of mistakes. Should be correct now.

No problem, it looks good now.

**Comment by greg0ire** [comment]
@sonata-project/contributors please review

File: Menu/MenuBuilder.php [hunk 3]
```diff
@@ -74,6 +74,7 @@ public function createSidebarMenu()
                 'icon' => $group['icon'],
                 'label_catalogue' => $group['label_catalogue'],
                 'roles' => $group['roles'],
+                'SonataAdminBundle' => true,
```
Comment (Koc) [inline Menu/MenuBuilder.php]: Maybe better use underscore for consistency?

**Comment by wouterj** [comment]
Updated to use underscores. Ready to merge

**Comment by greg0ire** [comment]
ping @sonata-project/contributors

**Comment by OskarStark** [comment]
What do you think to use `sonata_admin` instead of `sonata_admin_bundle` ?

**Comment by wouterj** [comment]
It's all fine by me. I don't care about the name of the setting, I only care about getting this fixed :)

**Comment by OskarStark** [comment]
> It's all fine by me. I don't care about the name of the attribute, I only care about getting this fixed :)

Thank you @wouterj 👍  Will merge after travis is green

**Comment by muspelheim** [comment]
@OskarStark ping

**Comment by greg0ire** [comment]
Thanks @wouterj !
